pyccl/p2d.py

```diff
@@ -0,0 +1,142 @@
+            try:
+                pkfunc(k=np.array([1E-2, 2E-2]), a=0.5)
+            except Exception:
+                raise ValueError("Can't use input function")
+
+            if cosmo is None:
+                raise ValueError("A cosmology is needed if initializing power spectrum from a function")
+
+            # Set k and a sampling from CCL parameters
+            nk = lib.get_pk_spline_nk(cosmo.cosmo)
+            na = lib.get_pk_spline_na(cosmo.cosmo)
+            a_arr, status = lib.get_pk_spline_a(cosmo.cosmo, na, status)
+            check(status)
+            lk_arr, status = lib.get_pk_spline_lk(cosmo.cosmo, nk, status)
+            check(status)
+
+            # Compute power spectrum on 2D grid
+            pkflat = np.zeros([na, nk])
+            for ia, a in enumerate(a_arr):
+                pkflat[ia, :] = pkfunc(k=np.exp(lk_arr), a=a)
+            pkflat = pkflat.flatten()
+
+        self.psp, status = lib.set_p2d_new_from_arrays(lk_arr, a_arr, pkflat,
+                                                       int(interp_order_lok),
+                                                       int(interp_order_hik),
+                                                       int(is_logp), status)
+        check(status)
+        self.has_psp = True
+
+    def eval(self, k, a, cosmo):
+        """Evaluate power spectrum.
+
+        Args:
+            k (float or array_like): wavenumber value(s) in units of Mpc^-1.
+            a (float): value of the scale factor
+            cosmo (:obj:`Cosmology`): Cosmology object. The cosmology object
+                  is needed in order to evaluate the power spectrum outside
+                  the interpolation range in `a`. E.g. if you want to evaluate
+                  the power spectrum at a very small a, not covered by the
+                  arrays you passed when initializing this object, the power
+                  spectrum will be extrapolated from the earliest available
+                  value using the linear growth factor (for which a cosmology
+                  is needed).
+            a_arr (array): an array holding values of the scale factor.
+
+        Returns:
+            float or array_like: value(s) of the power spectrum.
+        """
+        status = 0
+        cospass = cosmo.cosmo
+
+        if isinstance(k, int):
+            k = float(k)
+        if isinstance(k, float):
+            f, status = lib.p2d_eval_single(self.psp, np.log(k), a, cospass,
+                                            status)
+        else:
+            k_use=np.atleast_1d(k)
+            f, status = lib.p2d_eval_multi(self.psp, np.log(k_use), a, cospass,
+                                           k_use.size, status)
+        check(status, cosmo)
+
+        return f
+
+    def __del__(self):
+        """Free memory associated with this Pk2D structure
+        """
+        if hasattr(self, 'has_psp'):
+            if self.has_psp:
```

Comment:
you might want to check directly on psp argument too? I'm not sure.

Reply:
Fixed